src/Types/Type.php

```diff
@@ -137,14 +134,24 @@ public static function lookupName(self $type): string
     /**
      * Adds a custom type to the type map.
      *
-     * @param string             $name      The name of the type.
-     * @param class-string<Type> $className The class name of the custom type.
+     * @param string                  $name      The name of the type.
+     * @param class-string<Type>|Type $className The custom type or the class name of the custom type.
      *
      * @throws Exception
      */
-    public static function addType(string $name, string $className): void
+    public static function addType(string $name, string|Type $className): void
     {
-        self::getTypeRegistry()->register($name, new $className());
+        if ($className instanceof Type) {
+            self::getTypeRegistry()->register($name, $className);
+
+            return;
+        }
+
+        try {
+            self::getTypeRegistry()->register($name, new $className());
+        } catch (ArgumentCountError $e) {
+            throw TypeArgumentCountError::new($name, $e);
+        }
```

Comment:
We can wrap only the class instantiation into the try catch (after renaming the parameter to `$type`)

```suggestion
        if (is_string($type)) {
            try {
                $type = new $type();
            } catch (ArgumentCountError $e) {
                throw TypeArgumentCountError::new($name, $e);
            }
        }
        
        self::getTypeRegistry()->register($name, $type);
```